File: AISuite_QuickStart/app/src/main/java/com/zebra/aisuite_quickstart/java/analyzers/customdetector/yolo/YoloOnnxModel.java
// Copyright 2025 Zebra Technologies Corporation and/or its affiliates. All rights reserved.
package com.zebra.aisuite_quickstart.java.analyzers.customdetector.yolo;

import ai.onnxruntime.OnnxTensor;
import ai.onnxruntime.OrtEnvironment;
import ai.onnxruntime.OrtException;
import ai.onnxruntime.OrtSession;

import android.content.Context;
import android.graphics.Bitmap;
import android.graphics.Rect;
import android.util.Log;

import com.google.android.gms.tasks.Task;
import com.google.android.gms.tasks.Tasks;
import com.google.mlkit.vision.common.InputImage;
import com.zebra.ai.vision.entity.DetectionEntity;

import java.io.FileInputStream;
import java.io.IOException;
import java.nio.FloatBuffer;
import java.nio.MappedByteBuffer;
import java.nio.channels.FileChannel;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

/**
 * YOLOv8n general object detector — 80 COCO classes (ONNX Runtime).
 *
 * Input : [1, 3, 640, 640] float32 CHW, RGB normalised to [0, 1].
 * Output: [1, 84, 8400] — cx/cy/w/h + 80 class scores per prediction.
 *
 * OrtEnvironment is a JVM singleton — do NOT close it here; only the session is released.
 */
public class YoloOnnxModel implements AutoCloseable {

    private static final String TAG             = "YoloOnnxModel";
    private static final String MODEL_FILE      = "models/yolov8n-onnx/yolov8n.onnx";
    private static final int    INPUT_SIZE      = 640;
    private static final float  CONF_THRESHOLD  = 0.40f;
    private static final float  IOU_THRESHOLD   = 0.45f;
    private static final int    NUM_PREDICTIONS = 8400;
    private static final int    BOX_DIMS        = 4;
    private static final int    NUM_CLASSES     = 80;

    private final OrtEnvironment env;
    private final OrtSession     session;

    public YoloOnnxModel(Context context) throws IOException, OrtException {
        Log.d(TAG, "Loading ONNX session from assets: " + MODEL_FILE);
        env = OrtEnvironment.getEnvironment();
        android.content.res.AssetFileDescriptor fd = context.getAssets().openFd(MODEL_FILE);
        MappedByteBuffer buffer = new FileInputStream(fd.getFileDescriptor())
                .getChannel()
                .map(FileChannel.MapMode.READ_ONLY, fd.getStartOffset(), fd.getDeclaredLength());
        session = env.createSession(buffer);
        Log.i(TAG, "ONNX session ready — inputs=" + session.getInputNames()
                + " outputs=" + session.getOutputNames());
    }

    public Task<List<DetectionEntity>> process(
            InputImage inputImage) {
        try {
            List<DetectionEntity> result =
                    detect(inputImage.getBitmapInternal()).stream()
                            .map(rect -> new DetectionEntity(rect, new ArrayList<>()))
                            .collect(Collectors.toList());
            return Tasks.forResult(result);
        } catch (OrtException e) {
            return Tasks.forException(e);
        }
    }

    public List<Rect> detect(Bitmap bitmap) throws OrtException {
        float[] inputFloats = bitmapToInputFloats(bitmap);
        long[] shape        = {1L, 3L, INPUT_SIZE, INPUT_SIZE};
        String inputName    = session.getInputNames().iterator().next();

        OnnxTensor input = OnnxTensor.createTensor(env, FloatBuffer.wrap(inputFloats), shape);
        try {
            Map<String, OnnxTensor> inputMap = new HashMap<>();
            inputMap.put(inputName, input);
            try (OrtSession.Result results = session.run(inputMap)) {
                OnnxTensor outTensor = (OnnxTensor) results.get(0);
                FloatBuffer buf = outTensor.getFloatBuffer();
                float[] raw = new float[buf.remaining()];
                buf.get(raw);
                return decodeYolov8(raw, bitmap.getWidth(), bitmap.getHeight());
            }
        } finally {
            input.close();
        }
    }

    // YOLOv8n output: [1, 84, 8400] row-major — first index = feature, second = prediction
    // Channels 0-3: cx, cy, w, h in 640-px model space. Channels 4-83: 80 COCO class scores.
    private List<Rect> decodeYolov8(float[] raw, int bitmapW, int bitmapH) {
        List<float[]> boxes  = new ArrayList<>();
        List<Float>   scores = new ArrayList<>();

        for (int p = 0; p < NUM_PREDICTIONS; p++) {
            float maxScore = 0f;
            for (int c = 0; c < NUM_CLASSES; c++) {
                float s = raw[(BOX_DIMS + c) * NUM_PREDICTIONS + p];
                if (s > maxScore) maxScore = s;
            }
            if (maxScore < CONF_THRESHOLD) continue;

            float cx = raw[0 * NUM_PREDICTIONS + p];
            float cy = raw[1 * NUM_PREDICTIONS + p];
            float w  = raw[2 * NUM_PREDICTIONS + p];
            float h  = raw[3 * NUM_PREDICTIONS + p];
            boxes.add(new float[]{cx - w / 2f, cy - h / 2f, cx + w / 2f, cy + h / 2f});
            scores.add(maxScore);
        }

        if (boxes.isEmpty()) return new ArrayList<>();

        float[] scoresArr = new float[scores.size()];
        for (int i = 0; i < scores.size(); i++) scoresArr[i] = scores.get(i);

        List<Integer> kept = NmsUtils.nms(boxes, scoresArr, IOU_THRESHOLD);
        List<Rect> results = new ArrayList<>(kept.size());
        for (int idx : kept) {
            float[] b = boxes.get(idx);
            results.add(new Rect(
                    Math.round(b[0] / INPUT_SIZE * bitmapW),
                    Math.round(b[1] / INPUT_SIZE * bitmapH),
                    Math.round(b[2] / INPUT_SIZE * bitmapW),
                    Math.round(b[3] / INPUT_SIZE * bitmapH)));
        }
        return results;
    }

    private float[] bitmapToInputFloats(Bitmap bitmap) {
        Bitmap software = bitmap.getConfig() == Bitmap.Config.HARDWARE
                ? bitmap.copy(Bitmap.Config.ARGB_8888, false) : bitmap;
        Bitmap scaled = Bitmap.createScaledBitmap(software, INPUT_SIZE, INPUT_SIZE, true);
        int[] pixels = new int[INPUT_SIZE * INPUT_SIZE];
        scaled.getPixels(pixels, 0, INPUT_SIZE, 0, 0, INPUT_SIZE, INPUT_SIZE);

        int n = INPUT_SIZE * INPUT_SIZE;
        float[] floats = new float[3 * n];
        for (int i = 0; i < n; i++) {
            floats[i]         = (pixels[i] >> 16 & 0xFF) / 255f;
            floats[n + i]     = (pixels[i] >>  8 & 0xFF) / 255f;
            floats[2 * n + i] = (pixels[i]        & 0xFF) / 255f;
        }
        return floats;
    }

    @Override
    public void close() throws OrtException {
        session.close();
        // Do NOT close OrtEnvironment — it is a JVM-wide singleton
    }
}

File: AISuite_QuickStart/app/src/main/java/com/zebra/aisuite_quickstart/kotlin/analyzers/customdetector/yolo/YoloOnnxModel.kt
// Copyright 2025 Zebra Technologies Corporation and/or its affiliates. All rights reserved.
package com.zebra.aisuite_quickstart.kotlin.analyzers.customdetector.yolo

import ai.onnxruntime.OnnxTensor
import ai.onnxruntime.OrtEnvironment
import ai.onnxruntime.OrtSession
import android.content.Context
import android.graphics.Bitmap
import android.graphics.Rect
import android.util.Log
import java.io.FileInputStream
import java.nio.FloatBuffer
import java.nio.channels.FileChannel

/**
 * YOLOv8n general object detector — 80 COCO classes (ONNX Runtime).
 *
 * Input : [1, 3, 640, 640] float32 CHW, RGB normalised to [0, 1].
 * Output: [1, 84, 8400] — cx/cy/w/h + 80 class scores per prediction.
 *
 * Implements AutoCloseable so the ONNX session is released when
 * CustomDetector.close() cascades to this.
 */
class YoloOnnxModel(context: Context) : AutoCloseable {

    companion object {
        private const val MODEL_FILE      = "models/yolov8n-onnx/yolov8n.onnx"
        private const val INPUT_SIZE      = 640
        private const val CONF_THRESHOLD  = 0.3f
        private const val IOU_THRESHOLD   = 0.45f
        private const val TAG = "YoloOnnxModel"
    }

    private val env    : OrtEnvironment = OrtEnvironment.getEnvironment()
    private val session: OrtSession

    init {
        Log.d(TAG, "Loading ONNX session from assets: $MODEL_FILE")
        val fd     = context.assets.openFd(MODEL_FILE)
        val buffer = FileInputStream(fd.fileDescriptor).channel.map(
            FileChannel.MapMode.READ_ONLY, fd.startOffset, fd.declaredLength
        )
        session = env.createSession(buffer)
        Log.i(TAG, "ONNX session ready — inputs=${session.inputNames} outputs=${session.outputNames}")
    }

    fun detect(bitmap: Bitmap): List<Rect> {
        val inputFloats = bitmapToInputFloats(bitmap)
        val shape       = longArrayOf(1L, 3L, INPUT_SIZE.toLong(), INPUT_SIZE.toLong())
        val inputName   = session.inputNames.iterator().next()

        val raw = OnnxTensor.createTensor(env, FloatBuffer.wrap(inputFloats), shape).use { input ->
            session.run(mapOf(inputName to input)).use { results ->
                val outTensor = results[0] as OnnxTensor
                val buf = outTensor.floatBuffer
                FloatArray(buf.remaining()).also { buf.get(it) }
            }
        }

        return decodeYolov8(raw, bitmap.width, bitmap.height)
    }

    private fun bitmapToInputFloats(bitmap: Bitmap): FloatArray {
        val software = if (bitmap.config == Bitmap.Config.HARDWARE)
            bitmap.copy(Bitmap.Config.ARGB_8888, false) else bitmap
        val scaled = Bitmap.createScaledBitmap(software, INPUT_SIZE, INPUT_SIZE, true)
        val pixels = IntArray(INPUT_SIZE * INPUT_SIZE)
        scaled.getPixels(pixels, 0, INPUT_SIZE, 0, 0, INPUT_SIZE, INPUT_SIZE)

        val n = INPUT_SIZE * INPUT_SIZE
        val floats = FloatArray(3 * n)
        for (i in pixels.indices) {
            floats[i]         = (pixels[i] shr 16 and 0xFF) / 255f
            floats[n + i]     = (pixels[i] shr  8 and 0xFF) / 255f
            floats[2 * n + i] = (pixels[i]        and 0xFF) / 255f
        }
        return floats
    }

    // YOLOv8n output: [1, 84, 8400] row-major — first index = feature, second = prediction.
    // Channels 0-3: cx, cy, w, h in 640-px model space. Channels 4-83: 80 COCO class scores.
    private fun decodeYolov8(raw: FloatArray, bitmapW: Int, bitmapH: Int): List<Rect> {
        val numPreds  = 8400
        val numValues = 84

        val boxes  = mutableListOf<FloatArray>()
        val scores = mutableListOf<Float>()

        for (p in 0 until numPreds) {
            val cx = raw[0 * numPreds + p]
            val cy = raw[1 * numPreds + p]
            val w  = raw[2 * numPreds + p]
            val h  = raw[3 * numPreds + p]

            var maxScore = 0f
            for (c in 4 until numValues) {
                val s = raw[c * numPreds + p]
                if (s > maxScore) maxScore = s
            }
            if (maxScore < CONF_THRESHOLD) continue

            boxes  += floatArrayOf(
                (cx - w / 2f) * bitmapW / 640f,
                (cy - h / 2f) * bitmapH / 640f,
                (cx + w / 2f) * bitmapW / 640f,
                (cy + h / 2f) * bitmapH / 640f
            )
            scores += maxScore
        }

        return nms(boxes, scores.toFloatArray(), IOU_THRESHOLD)
            .map { i -> boxes[i].let { Rect(it[0].toInt(), it[1].toInt(), it[2].toInt(), it[3].toInt()) } }
    }

    // OrtEnvironment is a singleton — closing it would affect any other ONNX session.
    override fun close() = session.close()
}
